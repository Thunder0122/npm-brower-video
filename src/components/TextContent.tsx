import React from "react";

/**
 * Takes a string as children
 * Return ONE or MULTIPLE <p> tags split inside from the \n\n
 * - dir="auto" adapts ltr or rtl (right to left) languages (like perse/hebrew)
 * - text-align: initial to make sure to override `-webkit-match-parent` if set (which won't take language direction in account)
 */
const TextContent: React.FunctionComponent<
  { children: string } & React.HTMLAttributes<HTMLElement>
> = ({ children, style, ...props }) => (
  <>
    {children.split(`\n\n`).map((paragraph, index) => (
      <p
        dir="auto"
        style={{ textAlign: "initial", ...style }}
        {...props}
        key={index}
      >
        {paragraph}
      </p>
    ))}
  </>
);

export default TextContent;
