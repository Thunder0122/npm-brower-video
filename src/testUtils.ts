export {
  wait,
  waitForElement,
  render,
  cleanup,
  fireEvent
} from "react-testing-library";

// Add utils like `renderWithProvider()`
