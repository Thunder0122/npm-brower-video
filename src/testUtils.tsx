import React from "react";
import { render } from "react-testing-library";

import { appWithTranslationMock } from "./services/i18n/NamespaceMock";
import {
  LanguageManagerProvider,
  LanguageManagerConsumer
} from "./services/i18n/LanguageManager";
import { i18n as i18nInstance } from "../i18n";
import { setLanguageOverrideFromCookie } from "./services/i18n/utils";

/**
 * ⚠️ This is a hack, the other import ways don't work (how does `next-i18next` manages it internally ?... It's the same code ...):
 * Won't work:
 * - import { I18nextProvider } from "react-i18next";
 * - const I18nextProvider = require("react-i18next").I18nextProvider;
 */
const I18nextProvider = require("../node_modules/react-i18next/dist/commonjs/I18nextProvider")
  .I18nextProvider;

export {
  wait,
  waitForElement,
  render,
  cleanup,
  fireEvent
} from "react-testing-library";

/**
 * If your component is wrapped in `withNamespaces` and uses the `t` prop,
 * we need to mock it - see `src/services/i18n/NamespaceMock`
 * 
 * MAKE sure you mocked withNamespaces in the setup of the tests (setupTests.js):
 * 
```
const { withNamespacesMock } = require("./services/i18n/NamespaceMock");

jest.mock("react-i18next", () => ({
  withNamespaces: () => Component => Component
}));
```
 * @param ui 
 * @param renderOptions 
 */
export const renderI18nNamespacesWrappedComponent = (
  Comp: React.Component,
  renderOptions = {},
  { i18n = i18nInstance, defaultNS = "common", initialLanguage = "en" } = {}
) => {
  i18n.language = initialLanguage;
  const DecoratedWithTranslation = appWithTranslationMock(Comp, {
    i18n: i18nInstance,
    defaultNS,
    initialLanguage
  });
  return render(
    <I18nextProvider
      i18n={i18n}
      defaultNS={defaultNS}
      initialLanguage={initialLanguage}
    >
      <DecoratedWithTranslation />
    </I18nextProvider>,
    renderOptions
  );
};

/**
 * If your component is wrapped in `withNamespaces` and uses
 * the custom `LanguageManager.Consumer` inside, you'll need to render it with the following util
 * which will do the same as `renderI18nNamespacesWrappedComponent` and also
 * wrap your component with a `LanguageManager.Provider`, so you can access:
 * - language
 * - languageOverride
 * - languageOverrideFull
 * - switchLanguage
 *
 * It also wraps with a `LanguageManager.Consumer` (like in _app.tsx) in order to force re-render
 * when switchLanguage is called
 */
export const renderI18nWithLanguageManagerProvider = (
  Comp: typeof React.Component,
  renderOptions = {},
  {
    i18n = i18nInstance,
    defaultNS = "common",
    initialLanguage = "en",
    languageOverride = "en",
    languageOverrideFull = "en-US"
  } = {}
) => {
  i18n.language = initialLanguage;
  const App = () => (
    <I18nextProvider
      i18n={i18nInstance}
      defaultNS={defaultNS}
      initialLanguage={initialLanguage}
    >
      <LanguageManagerProvider
        i18n={i18nInstance}
        languageOverride={languageOverride}
        languageOverrideFull={languageOverrideFull}
      >
        <LanguageManagerConsumer>
          {languageManagerCtx => <Comp {...languageManagerCtx} />}
        </LanguageManagerConsumer>
      </LanguageManagerProvider>
    </I18nextProvider>
  );
  const AppWithTranslation = appWithTranslationMock(App, {
    i18n: i18nInstance,
    defaultNS,
    initialLanguage
  });
  return render(<AppWithTranslation />, renderOptions);
};

export const resetLanguage = ({
  initialLanguage = "en",
  languageOverride = "en",
  languageOverrideFull = "en-US"
} = {}) => () => {
  i18nInstance.language = initialLanguage;
  setLanguageOverrideFromCookie(languageOverride);
  setLanguageOverrideFromCookie(languageOverrideFull, true);
};
