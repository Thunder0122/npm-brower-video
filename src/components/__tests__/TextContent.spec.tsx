import TextContent from "../TextContent";

import { render } from "../../testUtils";

describe("src/components/TextContent", () => {
  it("should render ONE empty <p> from an empty string", () => {
    const { container } = render(<TextContent>{""}</TextContent>);
    expect(container.innerHTML).toMatchSnapshot();
  });
  it("should render ONE <p> from a string", () => {
    const { container } = render(<TextContent>{"Lorem ipsum"}</TextContent>);
    expect(container.innerHTML).toMatchSnapshot();
  });
  it("should render MULTIPLE <p> from a string containing carriage return", () => {
    const { container } = render(
      <TextContent>
        {"Lorem ipsum\n\ndolor sit amet\n\nconsectetur adipiscing elit"}
      </TextContent>
    );
    expect(container.innerHTML).toMatchSnapshot();
  });
  it("should spread props on each paragraph", () => {
    const { container } = render(
      <TextContent dir="auto" style={{ fontWeight: "bold" }}>
        {"Lorem ipsum\n\ndolor sit amet\n\nconsectetur adipiscing elit"}
      </TextContent>
    );
    expect(container.innerHTML).toMatchSnapshot();
  });
});
