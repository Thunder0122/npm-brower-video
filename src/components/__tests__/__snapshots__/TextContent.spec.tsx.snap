// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`src/components/TextContent should render MULTIPLE <p> from a string containing carriage return 1`] = `"<p dir=\\"auto\\" style=\\"text-align: initial;\\">Lorem ipsum</p><p dir=\\"auto\\" style=\\"text-align: initial;\\">dolor sit amet</p><p dir=\\"auto\\" style=\\"text-align: initial;\\">consectetur adipiscing elit</p>"`;

exports[`src/components/TextContent should render ONE <p> from a string 1`] = `"<p dir=\\"auto\\" style=\\"text-align: initial;\\">Lorem ipsum</p>"`;

exports[`src/components/TextContent should render ONE empty <p> from an empty string 1`] = `"<p dir=\\"auto\\" style=\\"text-align: initial;\\"></p>"`;

exports[`src/components/TextContent should spread props on each paragraph 1`] = `"<p dir=\\"auto\\" style=\\"text-align: initial; font-weight: bold;\\">Lorem ipsum</p><p dir=\\"auto\\" style=\\"text-align: initial; font-weight: bold;\\">dolor sit amet</p><p dir=\\"auto\\" style=\\"text-align: initial; font-weight: bold;\\">consectetur adipiscing elit</p>"`;
